torch/_dynamo/output_graph.py

```diff
@@ -961,6 +964,24 @@ def get_graph_sizes_log_str(self, name):
                         )
         return graph_sizes_str
 
+    @contextlib.contextmanager
+    def restore_global_state(self):
+        """
+        Momentarily restores the global state to what it was prior to tracing the current output
+        """
+        prior_global_state = self.tracing_context.global_context.copy_graphstate()
+        current_global_state: Dict[str, Tuple[Any, bool]] = {}
+        self.save_global_state(current_global_state)
+        try:
+            # Set to state prior to tracing the graph
+            self.tracing_context.global_context.restore_graphstate(prior_global_state)
+            yield
+        finally:
+            # Reset to state at the current time before calling the user compiler
+            self.tracing_context.global_context.restore_graphstate(
+                GlobalContextCheckpointState(current_global_state)
+            )
```

Comment:
This seems reasonable based on what we discussed but I want to see what @ezyang thinks.

Basically, the problem that this code is solving is that:

(1) When dynamo encounters a `torch.set_grad_enabled` call that it proxies into the graph, it **also** manually changes the global grad state ([code](https://github.com/pytorch/pytorch/blob/84f59d893a029ebb804ba344e48fb2c2c56a213d/torch/_dynamo/variables/ctx_manager.py#L160))

(2) Why does it do that? It needs the ambient grad state to be set properly so that it can trace code like this correctly:
```
        def fn4(a, b):
            x = a + 1
            with torch.set_grad_enabled(True):
                if torch.is_grad_enabled():
                    x = x + b
            x = x + 2
            return x
```

(3) In theory, instead of doing the above we could change dynamo to stop mutating the grad state during tracing. This would require some global VariableTracker representing all global state, that dynamo knows to update + query on any global state accesses during tracing.

John and I briefly discussed (3), and thought that it would be more involved than what's currently in this PR (just accepting that dynamo might have changed global state during tracing and snapping it back into place right before calling AOTAutograd)

Reply:
I don't know the history but I think part of it is just fake tensor tracing? The resultant faketensors will have the right requires_grad 

https://github.com/pytorch/pytorch/blob/68dead4a6cff1cdfd415ca9fdfc868ac135e6544/torch/_dynamo/variables/tensor.py#L103